---
title: "CITE-seq-count report"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, echo=FALSE}
knitr::opts_chunk$set(dev.args=list(png=list(type="cairo")))
```


```{r laod libraries, echo=FALSE, message=FALSE}
library(Seurat)
library(reshape)
library(ggplot2)
library(dplyr)
library(pheatmap)
```

## Overview

This document provides a basic report based on the output from CITE-seq-count which could have been run on antibody-derived tag (ADT) data coming from a multiplexed scRNAseq dataset or cite-seq dataset.

### Percentage mapped ADTs

Below is a plot that summarises the percentage of reads that have a mappable ADT according to the tag sequences that were specified in the run.

```{r percent mapped, echo=FALSE, message=FALSE, fig.height=5, fig.width=10}

pmapped <- read.csv("../run_report.dir/percentage_mapped.tsv", header=TRUE, stringsAsFactors=FALSE, sep="\t")
pmapped <- melt(pmapped)
ggplot(pmapped, aes(x=variable, y=value)) +
  geom_bar(stat="identity") +
  xlab("") +
  ylab("% mapped") + 
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90))
```

### Unmapped cell barcodes

It may be the case that there are problematic cells i.e those that do not have any truly mappable tag sequences. In this case, CITE-seq-count counts the number of unmapped tags and these would show up as a high proportion of unmapped counts for a given cell barcode. Below we try and get an idea of the number of unmapped vs mapped tag sequences for each cell barcode in each sample in order to flag any potentially problematic cells.

```{r unmapped counts, echo=FALSE, message=FALSE, fig.height=20, fig.width=10}

# function to do the plotting
plot_mapped <- function(data_dir){
  dat <- Read10X(data.dir=data_dir, gene.column=1)
  sample_name <- unlist(strsplit(data_dir, "/"))
  sample_name <- sample_name[3]
  dat <- as.data.frame(dat)
  
  dat$is_mapped <- ifelse(rownames(dat) == "unmapped", "no", "yes")
  dat$HTO <- rownames(dat)
  dat <- melt(dat)
  dat_mapped <- aggregate(value~variable + is_mapped, dat, sum)

  ggplot(dat_mapped, aes(x=variable, y=value, fill=is_mapped, group=is_mapped)) + 
    geom_bar(stat="identity") +
    scale_fill_manual(values=c("red", "blue")) +
    facet_wrap(~is_mapped) +
    ggtitle(sample_name) +
    theme(axis.text.x=element_blank(),
          axis.title.x=element_blank()) +
    ylab("count") +
    xlab("barcode")
}

grobs.list <- list()
dirs <- list.dirs("../cite-seq-count.dir/", recursive=FALSE)
for (i in 1:length(dirs)){
    dir <- paste0(dirs[i], "/", "read_count")
    p <- plot_mapped(dir)
    grobs.list[[i]] <- p
}
gridExtra::grid.arrange(grobs=grobs.list, ncol=1)

```

### Counts for each ADT across samples

In theory there shouldn't be a bias in counts across ADTs in a multiplexed experiment (this would be different in a cite-seq project). Here we look at the counts assigned to eah ADT across samples. Each ADT should have counts that are quite similar across runs (if there are multiple runs).

```{r ADT counts across samples, echo=FALSE, message=FALSE, fig.height=5, fig.width=10}

 get_adt_counts <- function(data_dir){
  
  dat <- Read10X(data.dir=data_dir, gene.column=1)
  sample_name <- unlist(strsplit(data_dir, "/"))
  sample_name <- sample_name[4]
  dat <- as.data.frame(dat)
  dat <- dat[rownames(dat) != "unmapped",]
  # take the sum across cells for each ADT
  sums <- data.frame(rowSums(dat))
  
  # should normalise for the total counts
  rownames(sums) <- rownames(dat)
  total <- sum(sums[,1])
  sums <- sums/total
  colnames(sums) <- sample_name
  return(sums)
  }

adt_counts <- list()
dirs <- list.dirs("../cite-seq-count.dir/", recursive=FALSE)
for (i in 1:length(dirs)){
    dir <- paste0(dirs[i], "/", "read_count")
    res <- get_adt_counts(dir)
    ADTs <- unlist(strsplit(rownames(res), "-"))
    ADTs <- ADTs[seq(2, length(ADTs), 2)]
    rownames(res) <- ADTs
    adt_counts[[i]] <- res
}
adt_counts <- data.frame(dplyr::bind_cols(adt_counts))
rownames(adt_counts) <- rownames(res)

adt_counts$ADT <- rownames(res)
adt_counts <- melt(adt_counts)
colors <- ggsci::pal_simpsons()(length(unique(adt_counts$variable)))
ggplot(adt_counts, aes(x=ADT, y=value, color=variable)) +
#  geom_boxplot() +
  geom_jitter(width=0.15) +
  theme_bw() +
  coord_flip() +
  xlab("ADT sequence") +
  ylab("Fraction of mapped reads (sum across cells)") +
  scale_color_manual(values=colors)
```





